packages/dds/SchemaVersioning.md

```diff
@@ -0,0 +1,159 @@
+
+1.  Any changes in document schema must ship dark (default to not being used for writing) and be roll out (enabling new format) on a schedule controlled by the controlled the team controlling the application deployment schedule.
+
+    Writing new formats can only be enabled by application only when application knows it is close to 100% of clients having new bits to collaborate in new schema (or it’s Ok for old clients to fail in some controlled way).
+
+1.  When a new format is added, support for writing the previous format must available in the same version that has support for the new format.
+
+# Best Practices
+
+Here are some patterns to follow to meet the above requirements while optimizing for maintainability and user experience.
+
+## General
+
+Both libraries and applications can use these.
+
+-   Update Data to new formats lazily (when the data would be rewritten anyway).
+    This approach as several benefits over eager updating:
+
+    1. Works with no special handling for readonly clients.
+    2. Avoids new formats causing write amplification.
+    3. Keeps providence metadata (like who edited the content most recently) accurate.
+    4. Reduces the impact of compatibility issues. For example just opening a document in one client will never break it for another. This also reduces the amount of damage caused by a bad update before it can be rolled back.
+
+-   Always check for unsupported formats, and report recoverable errors including what the format is that is not supported.
+    Libraries should make it as easy as possible for applications to expose this error to the users so they can ideally update the application and continue without data-loss.
+    However the case where updating the application does not solve the issue should be handled as well: ensure no data corruption can occur due to unexpected formats being encountered as well as that if data corruption did occur and made a format appear unrecognized, that telemetry could detect the issue.
+
+-   Keep the contents of persisted data formats out of public APIs.
+
+    Since APIs have different compatibility requirements than persisted data formats, mixing the two will result in reduced evolvability long term.
+    This avoids the complexity related to supporting reading and writing multiple formats from cluttering the APIs.
+
+-   Implement support for persisted formats in a self contained way, where the relevant code is clearly marked.
+
+    A type which is persisted has very different evolvability than one that is not.
+    For example a package that has a non-exported interface usually can do a rename refactoring for members of the interface, but if that interface is used for Json serialized objects which are persisted,
+    the change could easily break support for all existing documents and/or collaboration with all existing clients, and do so in a way that won't show up in code review since the persistance/format logic might be untouched.
```

Comment:
```suggestion
    the change could easily break support for all existing documents and/or collaboration with all existing clients, and do so in a way that won't show up in code review since the persistence/format logic might be untouched.
```